text/0314-cdk-construct-hub.md

```diff
@@ -0,0 +1,626 @@
+help developers integrate those constructs into their application.
+
+Customers can also deploy a private instance of Construct Hub to facilitate
+sharing infrastructure abstractions published to a private package registry
+within the organization.
+
+### Where can I find the Construct Hub?
+
+The public instance of Construct Hub is at: `${TBD}`.
+
+### How do I publish a package to Construct Hub?
+
+Packages published to the public npm registry (npmjs.com) are automatically
+indexed by the public instance of Construct Hub as long as they meet the
+following prerequisites:
+
+1. Cross-language support has been configured correctly using [jsii]
+1. The `cdk` keyword is referenced (in the `keywords` key of `package.json`)
+
+[jsii]: https://aws.github.io/jsii
+
+### Can I opt packages out of the public instance of Construct Hub?
+
+There is currently no self-service way to un-list a package from the Construct
+Hub. If you would like your package to be removed from the public instance of
+Construct Hub, please file an issue in GitHub. You will be required to prove
+ownership of the package(s) you are requesting un-listing for.
+
+### Can I trust all constructs listed on Construct Hub?
+
+Construct Hub does not perform a comprehensive audit of packages it indexes.
+You should not trust packages found via the Construct Hub any differently
+from packages found via other package search tools or registries, such as
+npmjs.com, Maven Central, PyPI, NuGet, etc.
+
+## Specification
+
+### Landing Page
+
+> [Example rendering](#rendering-landing-page)
+
+The landing page for the Construct Hub includes a search bar that allows
+developers to search for constructs that solve their use-case using relevant
+keywords, and optionally filtering on a specific programming language and
+constructs domain (AWS CDK / CDK8s / CDKtf / Projen / ...).
+
+It may also feature a short-list of packages, selected based on popularity or
+latest release date.
+
+### Search Results Page
+
+When a search is performed using the search bar in the landing page (or in the
+header of all other pages), a list of relevant results is displayed as tiles,
+similar to the short-list of packages shown on the landing page.
+
+### Package Detail Page
+
+> [Example rendering](#rendering-package-detail-page)
+
+Clicking on a given package's tile in the landing page or search result page
+brings the user to a package detail page that represents the following
+information:
+
+- Package name, tags, and description
+- Overview of publishing information (author, version, license, etc...)
+  - Where possible, links to the source repository, issue tracker, etc... based
+    on information registered in the `.jsii` assembly file.
+- Install instructions
+- Available languages
+- Package's documentation
+  - Rendered content of the package's `README.md`
+  - API reference documentation, in all supported languages (only the version
+    for the currently selected language is displayed)
```

Comment:
Another roadmap idea: "virtual construct libraries" for L1s. For example, one idea that came up is to automatically generate a virtual construct library for each Terraform provider from the Terraform registry and publish them to the hub with instructions on how to use `cdktf import` in order to generate their constructs. Same for k8s CRDs and CFN registry resources. This might be a great way to seed the hub with tons of great and useful content.

Reply:
> A feature we should add to the readme is an "Edit" button in a package overview page and jump directly to the GitHub repository with that README.md file to propose an edit (bonus points if we can do something similar in reference docs, but that's borderline rocket science).

I don't think those are V1/MVP features. They also sound like wee UI elements that can be easily done later on.

---

> Another roadmap idea: "virtual construct libraries" for L1s. For example, one idea that came up is to automatically generate a virtual construct library for each Terraform provider from the Terraform registry and publish them to the hub with instructions on how to use `cdktf import` in order to generate their constructs. Same for k8s CRDs and CFN registry resources. This might be a great way to seed the hub with tons of great and useful content.

That sounds like a whole new set of features, with a very different source and data flow than what is currently discussed here. While that sounds like an interesting idea, I don't think it can fit in our aspirational timeline.

Reply:
Both of these ideas are roadmap (I.e future ideas). I don't expect us to support them in v1 but it's sometimes useful to list some future directions in the initial spec to inform the design.